### PR TITLE
Improve the way QType behaves by defining proper conversion operators and a hash function.

### PR DESCRIPTION
**Pull request by omoerbeek**
I converted syncres.* for starters to stop using uint16_t for qtype
and use QType everywhere.  Note that I also changed const QType&
in arg lists by QType, since it makes little sense to pass a 16 bit value
by const reference.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/qtype.cc [hunk 1]
```diff
@@ -30,23 +30,79 @@
 
 static_assert(sizeof(QType) == 2, "QType is not 2 bytes in size, something is wrong!");
 
-vector<QType::namenum> QType::names;
-// XXX FIXME we need to do something with initializer order here!
-QType::init QType::initializer; 
-
-QType::QType()
+const vector<QType::namenum> QType::names = {
```
Comment (pieterlexis) [inline pdns/qtype.cc:33]: I wonder if we could wrap this into `boilerplate_conv` via DNSRecords. Or make `boilerplate_conv` take a QType, so we never have the problem of types not being in sync with supported types?

Comment (omoerbeek) [reply pdns/qtype.cc:33]: I changed the boilerplate so that it uses QType::RNAME, enforcing it must exists. That's a start, I'd say.

**Comment by pieterlexis** [comment]
> Note that I also changed const QType& in arg lists by QType, since it makes little sense to pass a 16 bit value by const reference.

It might still be good to keep them `const` just in case?

**Comment by omoerbeek** [comment]
> > Note that I also changed const QType& in arg lists by QType, since it makes little sense to pass a 16 bit value by const reference.
> 
> It might still be good to keep them `const` just in case?

It is uncommon the specify `const` in a prototype non-reference or pointer args.

It's call by value anyway so the caller is safe form modifications by the callee.  As for the callee, `const` only prevents changing the value in its body.

File: pdns/qtype.hh [hunk 2]
```diff
@@ -38,206 +36,123 @@
...
 
-
-
 class QType
 {
 public:
-  QType(); //!< Naked constructor
-  explicit QType(uint16_t); //!< convert from an integer to a QType
-  QType(const QType& orig) : code(orig.code)
+  QType(uint16_t qtype = 0) : code(qtype) {}
```
Comment (rgacogne) [inline pdns/qtype.hh:42]: I'm not sure it's such a good idea to remove the `explicit` keyword on the constructor from `uint16_t`?

**Comment by rgacogne** [comment]
I agree it does not make sense to mark the `QType` as `const` in the declaration, but it might still make sense to mark it as `const` in the definition, as it does prevent mistakenly changing the value in the function.

File: pdns/qtype.hh (hunk 2 above)
Comment (omoerbeek) [reply pdns/qtype.hh:42]: It's conventient to be able to pass e.g. `QType::A` to a function taking a `QType`.

Comment (rgacogne) [reply pdns/qtype.hh:42]: That makes sense, so let's keep it that way.

**Comment by pieterlexis** [comment]
@omoerbeek can you rebase so we can review?

**Comment by omoerbeek** [comment]
Rebased

**Review by pieterlexis — APPROVED** [review]
:heart:

**Review by Habbie — APPROVED** [review]

**Pull request merged**